### ISSUE TITLE
(MariaDB/Mysql) Deadlock found when trying to get lock; try restarting transaction

### ISSUE THREAD

**Issue by erkexzcx** [openobserve/openobserve#4946]
Hey. I have 3 nodes cluster, each has these components:

1. OpenObserve cluster (0.13.0 currently, but happened on 0.12.* as well, I don't recall if previous versions had this issue)
2. ETCD cluster
3. Minio S3 cluster
4. MariaDB galera cluster

Some time ago I recall cluster worked great, but recently, with some previous versions (can't recall via which version it started happening), OpenObserve stopped loading data in webui - after long loading, it times out with an error... Restarting all nodes of OpenObserve fixes this issue and it works again for a while, but over time it happens again:

Logs from OpenObserve, with `warn` rust log level:
```
2024-10-29T12:00:51.198631245+00:00 ERROR infra::file_list::mysql: [MYSQL] commit select file_list_jobs pending jobs error: error returned from database: 1213 (40001): Deadlock found when trying to get lock; try restarting transaction    
2024-10-29T12:00:51.198685199+00:00 ERROR openobserve::job::compactor: [COMPACTOR] run data merge error: SqlxError# error returned from database: 1213 (40001): Deadlock found when trying to get lock; try restarting transaction
```

Any tips on how to fix this? Or what might be causing this?

EDIT:

Env variables that I use:
```
      ZO_LOCAL_MODE: false
      ZO_INSTANCE_NAME: ...
      ZO_ROOT_USER_EMAIL: ...
      ZO_ROOT_USER_PASSWORD: ...
      ZO_ETCD_ADDR: 127.0.0.1:2379
      ZO_META_STORE: mysql
      ZO_META_MYSQL_DSN: mysql://asdf:asdf@127.0.0.1:3306/openobserve
      ZO_S3_SERVER_URL: http://127.0.0.1:9000
      ZO_S3_REGION_NAME: us-east-1
      ZO_S3_ACCESS_KEY: ...
      ZO_S3_SECRET_KEY: ...
      ZO_S3_BUCKET_NAME: openobserve
      ZO_S3_PROVIDER: minio
      ZO_DATA_DIR: /var/lib/openobserve/
      ZO_INGEST_ALLOWED_UPTO: 168
      ZO_TELEMETRY: false
      ZO_MMDB_DISABLE_DOWNLOAD: true
      RUST_LOG: warn
```

**Comment by hengfeiyang** [comment]
Are you using `v0.13.0`? and How many compactors do you have? or you run it as a single node?

**Comment by erkexzcx** [comment]
Hey @hengfeiyang,

> Are you using v0.13.0?

Yes, but it also kept happening with `0.12.*`.

> How many compactors do you have?

I assume 3, because it's 3 nodes cluster and I did not set any separate roles for any node, so all 3 uses defaults and I assume there are 3 compactors?

> or you run it as a single node?

No, it's 3 nodes cluster.

-----

By the way, I've experimented a bit and I think I managed to fix this:

1. Set `ZO_META_TRANSACTION_LOCK_TIMEOUT` env variable from defaults (10m) to 60s. Looks like this mitigates this - if certain http requests fail in browser, then I need to wait 1 minute and re-try and deadlocks shouldn't be there.
2. **It looks like I completely fixed it** by switching Caddy load balancer, from `lb_policy least_conn` to `lb_policy client_ip_hash` load balancing method (aka enabled _sticky sessions_) and no more timing out in webui.

Let me come back here tomorrow or next week to confirm if this is indeed fixed. If it is - then you will know that users who use cluster version should use sticky sessions for load balancers. No idea how this fixes this issue, but for the past ~2 hours it did not happen anymore.
